File: src/main/java/com/derongan/minecraft/deeperworld/world/Region.java
package com.derongan.minecraft.deeperworld.world;

/**
 * Represents a region of the world. Contains all Y values.
 */
public class Region {
    private Point a;
    private Point b;

    public Region(int ax, int az, int bx, int bz) {
        a = new Point(ax,az);
        b = new Point(bx, bz);
    }

    public boolean contains(int x, int z) {
        return x <= Math.max(a.getX(), b.getX()) && x >= Math.min(a.getX(), b.getX()) && z <= Math.max(a.getZ(), b.getZ()) && z >= Math.min(a.getZ(), b.getZ());
    }


    public boolean contains(Point p) {
        return contains(p.getX(), p.getZ());
    }

    public Point getA() {
        return a;
    }

    public Point getB() {
        return b;
    }

    public Point midPoint(){
        return a.plus(b).div(2);
    }
}

File: src/main/kotlin/com/derongan/minecraft/deeperworld/world/Region.kt
package com.derongan.minecraft.deeperworld.world

import kotlinx.serialization.KSerializer
import kotlinx.serialization.Serializable
import kotlinx.serialization.builtins.ListSerializer
import kotlinx.serialization.builtins.serializer
import kotlinx.serialization.descriptors.SerialDescriptor
import kotlinx.serialization.encoding.Decoder
import kotlinx.serialization.encoding.Encoder
import kotlin.math.max
import kotlin.math.min

/**
 * Represents a region of the world.
 */
@Serializable(with = RegionSerializer::class)
class Region(val a: CubePoint, val b: CubePoint) {
    constructor(ax: Int, ay: Int, az: Int, bx: Int, by: Int, bz: Int) : this(
        CubePoint(ax, ay, az),
        CubePoint(bx, by, bz)
    )

    val center: CubePoint get() = a.plus(b).div(2)

    fun contains(x: Int, y: Int, z: Int): Boolean = x in min(a.x, b.x)..max(a.x, b.x) &&
            y in min(a.y, b.y)..max(a.y, b.y) &&
            z in min(a.z, b.z)..max(a.z, b.z)

    operator fun contains(p: CubePoint) = contains(p.x, p.y, p.z)
}

object RegionSerializer : KSerializer<Region> {
    private val serializer = ListSerializer(Int.serializer())
    override val descriptor: SerialDescriptor = serializer.descriptor

    override fun serialize(encoder: Encoder, value: Region) {
        val (x, y, z) = value.a
        val (x2, y2, z2) = value.b
        encoder.encodeSerializableValue(serializer, listOf(x, y, z, x2, y2, z2))
    }

    override fun deserialize(decoder: Decoder): Region {
        val (x, y, z, x2, y2, z2) = decoder.decodeSerializableValue(serializer)
        return Region(x, y, z, x2, y2, z2)
    }
}

private operator fun <E> List<E>.component6(): E = this[5]


